aten/src/ATen/native/TensorIterator.h

```diff
@@ -286,6 +287,18 @@ struct CAFFE2_API TensorIterator {
   /// CUDA reductions.
   bool is_final_output() const { return final_output_; }
 
+  bool has_promotion() const {
```

Comment:
can you please have a set_promotion_ field that would be set during compute_common_type? Additional iteration over operands calling scalar_type causes overhead.

Reply:
@ngimel I don't think there is an easy solution for this:

The common dtype is computed by coping all the tensors to a vector and invoke `at::native::result_type(tensorlist)`. The actual computation is done inside there, and the `compute_common_type` just wraps that function. It is hard to tell whether inside `result_type` there is type promotion or all the types are the same, unless we hack `result_type`. As we can see from part of the `compute_common_type`:
```C++
  //TODO refactor so that no tensor copies are done
  std::vector<Tensor> tensors;
  std::transform(std::begin(operands), std::end(operands), std::back_inserter(tensors),
                  [](const OperandInfo& op) { return op.tensor; });
  auto dtype = at::native::result_type(tensors);
  auto result = std::make_tuple(device, dtype, false);
  TORCH_INTERNAL_ASSERT(dtype != ScalarType::Undefined);
  return result;
```

If we just set the `has_promotion_` field unconditionally inside `compute_common_type`, then the `has_promotion()` will always return true, unless we call `dont_compute_common_dtype()`. This means, all the GPU loops would have a dynamic casting, even if all the dtypes are the same.

Reply:
[I deleted my comment, it seems to be wrong]